### DISCUSSION TITLE
Simple REST template

### DISCUSSION THREAD

**Discussion by Thihup** [piranhacloud/piranha#2582] [Show and tell]
I have created a simple project using the Piranha Embedded.
I tried to make CDI work in most places. I'd like some feedback on it :D

https://github.com/Thihup/piranha-rest-template/

There are a few issues with some dependencies, but I have already opened the issues to the respective projects.

**Comment by mnriem** [comment]
Have you seen the Jersey test in test/jersey?

**Reply by Thihup** [reply]
Sure :) 
With this template, I was trying to build an application from scratch as a user would do ;)

**Reply by mnriem** [reply]
Right, I just wonder if moving the Java test file into the src/main/java directory and changing what needs to be changed would work for you to get it to work with Jersey?

**Reply by Thihup** [reply]
I actually don't know. I wanted to try out RestEasy and it worked. I thought I'd have a hard time setting up CDI and Jersey, because of a few bad past experiences, but I just changed the dependencies and it worked too.

**Reply by Thihup** [reply]
(and with this experiment I could test that Piranha actually works with RestEasy and also reported an issue to them too 😉 )

**Reply by mnriem** [reply]
Should we add it to the piranhacloud organization as another template project. Maybe call it piranha-embedded-resteasy ?

**Reply by Thihup** [reply]
I've replaced RestEasy with Jersey due to some split package issues, maybe in the future 😅

**Reply by mnriem** [reply]
OK cool, once you have it ready let me know and I'll create the piranha-embedded-rest template project?

**Comment by OndroMih** [comment]
I think it would be useful to create a Jersey/RestEasy extension to abstract all the boilerplate, wouldn't it? I had a look at both the Jersey test and Thiago's rest template, and they both contain some unnecessary boilerplate (Jersey test uses web.xml and and the rest template uses a CDI extension).

I'll try to create a basic Jersey extension and see how it goes.

**Reply by Thihup** [reply]
Related to CDI: as we don't have an annotation processor in the Embedded, I used CDI as a way to get the classes.

Also, using CDI made everything be AtInjectable, something very similar to Spring.

Maybe "Simple" would not be the correct term here, but as Jakarta REST will require CDI by default in the future, including it now is a preparation for the future

**Reply by OndroMih** [reply]
I was thinking about annotations in Embedded and it looks like it's enough to add the AnnotationScanExtension before loading a custom extension to be able to use the annotationManager to read the annotations. But it's a manual task to do by developers and it's not documented anywhere. Have you thought about making it easier to compose extensions into a runtime? For example, specify dependencies between annotations (e.g. Jersey annotation would require either AnnotationScanExtension directly or would require an extension that "provides annotation manager").

I kind of like the way how Quarkus steps in during the build process (e.g. using the Quarkus maven plugin) and generates the runtime according to the code. If we use that approach, a maven plugin could scan the application, detect which extensions are needed, download their maven artifacts if they aren't defined in the pom.xml already, and build a custom launcher that initializes the Embedded Piranha with all the extensions. This would allow developers to just add the maven plugin and required Jakarta EE and MicroProfile APIs and all the rest could be handled by the plugin, making it even easier to bootstrap a new project for evelopers than with Quarkus.

Right now, I'm aiming to create a Jersey extension that will build on a few extensions and will require that those extensions are added to Piranha Embedded. Everything should be handled by that extension, without additional boiler-plate, so that developers only need to write the REST resources.
